core/chapters/c06_lists.py

```diff
@@ -420,95 +420,277 @@ class index_error(Step):
         def check(self):
             return "IndexError" in self.result
 
-    class introducing_len_and_range(VerbatimStep):
+    class improve_with_list_and_loop(VerbatimStep):
         """
 There you go. `words[4]` and beyond don't exist, so trying that will give you an error.
+That first program is a bit repetitive. Let's improve it with a list and a loop!
 
-By the way, you can get the number of elements in a list (commonly called the *length*) using `len(words)`.
-That means that the last valid index of the list is `len(words) - 1`, so the last element is `words[len(words) - 1]`. Try these for yourself.
+__program_indented__
+        """
 
-So in general, the valid indices are:
+        auto_translate_program = False
 
-    [0, 1, 2, ..., len(words) - 2, len(words) - 1]
+        def program(self):
+            words = ['This', 'is', 'a', 'list']
+            indices = [0, 1, 2, 3]
 
-There's a handy built in function called `range` which gives you these values. Try it by running this code:
+            for index in indices:
+                print(index)
+                print(words[index])
 
-__program_indented__
+    class using_range_instead_of_indices(VerbatimStep):
+        """
+That's a bit better, but writing out `[0, 1, 2, ...]` isn't great, especially if it gets long.
+There's a handy function `range` to do that part for you. Replace `[0, 1, 2, 3]` with `range(4)`,
+i.e. `indices = range(4)`.
         """
 
+        program_in_text = False
+
+        auto_translate_program = False
+
         def program(self):
-            for i in range(10):
-                print(i)
+            words = ['This', 'is', 'a', 'list']
+            indices = range(4)
+
+            for index in indices:
+                print(index)
+                print(words[index])
 
-    class range_len(VerbatimStep):
+    class printing_the_range(VerbatimStep):
         """
-`range(n)` is similar to the list `[0, 1, 2, ..., n - 2, n - 1]`.
-This gives us an alternative way to loop over a list:
+As you can see, the result is the same. Try this:
 
-__program_indented__
+    __copyable__
+    __program_indented__
         """
 
         predicted_output_choices = ["""\
-This
-is
-a
-list
-""", """\
 0
 1
 2
 3
 """, """\
-0
-This
 1
-is
 2
```

Comment:
Let's keep these predicted choices on `introducing_len_and_range` (currently `improve_with_list_and_loop`)

Reply:
adding them back...

Reply:
Is this OK?
```python
        predicted_output_choices = ["""\
0
1
2
3
This
is
a
list
""",  """\
1
2
3
4
This
is
a
list
""", """\
This
is
a
list
0
1
2
3
""", """\
0
This
1
is
2
a
3
list
""", """\
1
This
2
is
3
a
4
list
""",                                     ]
```

Reply:
No, use the same choices as before.